swarm_copy_tests/test_agent_routine.py

```diff
@@ -0,0 +1,563 @@
+                            },
+                        },
+                    }
+                ],
+                "tool_choice": None,
+                "stream": False,
+                "parallel_tool_calls": True,
+            }
+        )
+
+        # assert response content
+        assert response.choices[0].message.role == "assistant"
+        assert response.choices[0].message.content == "sample response content"
+
+    def test_handle_function_result(self, mock_openai_client):
+        routine = AgentsRoutine(client=mock_openai_client)
+
+        # Raw result is already a result
+        raw_result = Result(value="Nice weather")
+        result = routine.handle_function_result(raw_result)
+        assert result == raw_result
+
+        # Raw result is an agent for handoff
+        raw_result = Agent(name="Test agent 2")
+        result = routine.handle_function_result(raw_result)
+        assert result == Result(
+            value=json.dumps({"assistant": raw_result.name}), agent=raw_result
+        )
+
+        # Raw result is a tool output (Typically dict/list dict)
+        raw_result = [{"result_1": "Great result", "result_2": "Bad result"}]
+        result = routine.handle_function_result(raw_result)
+        assert result == Result(value=str(raw_result))
+
+    @pytest.mark.asyncio
+    async def test_execute_tool_calls(
+        self, mock_openai_client, get_weather_tool, agent_handoff_tool
+    ):
+        routine = AgentsRoutine(client=mock_openai_client)
+
+        # Test tool execution with simple tool call
+        mock_openai_client.set_response(
+            create_mock_response(
+                message={"role": "assistant", "content": ""},
+                function_calls=[
+                    {"name": "get_weather", "args": {"location": "Geneva"}}
+                ],
+            ),
+        )
+        agent = Agent(tools=[get_weather_tool, agent_handoff_tool])
+        context_variables = {}
+
+        tool_call_message = await routine.get_chat_completion(
+            agent,
+            history=[{"role": "user", "content": "Hello"}],
+            context_variables=context_variables,
+            model_override=None,
+        )
+        tool_calls = tool_call_message.choices[0].message.tool_calls
+        tool_calls_result = await routine.execute_tool_calls(
+            tool_calls=tool_calls,
+            tools=agent.tools,
+            context_variables=context_variables,
+        )
+        assert isinstance(tool_calls_result, Response)
+        assert tool_calls_result.messages == [
+            {
+                "role": "tool",
+                "tool_call_id": tool_calls[0].id,
+                "tool_name": "get_weather",
+                "content": "It's sunny today.",
+            }
+        ]
+        assert tool_calls_result.agent is None
+        assert tool_calls_result.context_variables == context_variables
+
+        # Test tool execution with multiple tool call + context_variables
```

Comment:
I have a general suggestion. I think it is better to separate test cases into different functions. This way you see easier what functionality works and what doesn't. Right now if a big test fails it does not tell you which case works and which case does not.